Add xorg sources to CI

matt335672 opened the pull request:
Created mainly for discussion at this point. I've got to a point where I need some input to make this workable.

This PR has been prompted by #340 and #362.

The PR adds a script which runs as part of the standard CI to make a basic build of the xorgxrdp module against the latest master from https://gitlab.freedesktop.org/xorg/xserver

The script has been tested on Ubuntu 22.04, 24.04 and Debian testing

I've got a couple of questions which I'd like anyone (including @metux) to comment on:-
1) Is the checkin CI the right place for this, or should we be doing a nightly or weekly run, like the Coverity scan for xrdp? 
2) Is a basic build OK, or should we be going for as comprehensive a build as possible?

Personally I'm thinking a nightly or weekly run would be better, but a checkin is fine, in which case I need to add some caching to this.

matt335672 committed:
Add script to build latest xorg from source

scripts/build_latest_xorg_apt.sh fetches the latest master source
for the xorg server and builds it.

matt335672 committed:
Add build against Xorg to checkin CI

matt335672 committed:
Fix shellcheck warnings

metux commented on scripts/build_latest_xorg_apt.sh:
```diff
@@ -0,0 +1,231 @@
+SOURCE_DIR="$HOME/xserver-src"
+
+# Dependencies in the source dir
+BUILDER="$SOURCE_DIR/util/modular/build.sh" ; # xorg modular build script
```
Where is this script coming from ?

matt335672 replied:
It's fetched here:-

https://github.com/neutrinolabs/xorgxrdp/blob/4e9b011e9bcff8affb77e65bb6a9e8cdc92cde1b/scripts/build_latest_xorg_apt.sh#L210-L213

This could be clearer though. I'll update the comment and add an additional check after the checkout.

metux commented:
> 1. Is the checkin CI the right place for this, or should we be doing a nightly or weekly run, like the Coverity scan for xrdp?

Since you wanna catch possibly breaking changes in Xorg upstream, a cyclic (eg. nightly) run seems plausible. Otherwise you'd only recognize problems when somebody creates a MR.

A good place for catching *new* things (that aren't in master yet), you can additionally test against my incubator branch:
https://gitlab.freedesktop.org/metux/xserver/-/tree/incubate?ref_type=heads

But be aware that this branch is unstable (and frequently rebased), so your test might create certain false alarms.

matt335672 commented:
Thanks @metux 

I can look at using your incubator, certainly. Is there a corresponding util/modular repo (or branch) I should be using with it?

matt335672 committed:
build_latest_xorg_apt.sh : Address review comments

matt335672 committed:
Move latest xorg CI build to a separate section

The CI build has been moved to a separate section to make
the changes in logic clearer.

Also:-
- Build with Glamor
- Run tests
- add caching of the xorg server build output so developers
  working on a feature can save time

matt335672 commented:
I've added two extra commits to make the script a bit clearer, and also to make these changes to the CI:-
1) Xorg latest build is in a separate section
2) Xorg latest build now also uses Glamor
3) Xorg latest build is run up briefly with xorgxrdp module.
3) Latest build directory is cached for a particular branch to speed up CI for developers.

I think this is fine for regular commit CI. We can also add a daily/weekly CI against @metux's inubator as a separate PR.

metux commented:
> Is there a corresponding util/modular repo (or branch)

I don't think anybody at xorg is still using these ancient scripts.

matt335672 commented:
I see!

I picked them up as it's the first option here:-

https://www.x.org/wiki/Building_the_X_Window_System/

Is there a better, more maintainable way to accomplish an X server build you would recommend?
